// @tarojs/components/dist-h5/vue3/components/index.js
import createComponent from '../createComponent';
import createFormsComponent from '../createFormsComponent'; // 视图容器

export var CoverImage = createComponent('taro-cover-image');
export var CoverView = createComponent('taro-cover-view');
export var MatchMedia = createComponent('taro-match-media');
export var MovableArea = createComponent('taro-movable-area');
export var MovableView = createComponent('taro-movable-view');
export var PageContainer = createComponent('taro-page-container');
export var RootPortal = createComponent('taro-root-portal');
export { default as ScrollView } from './scroll-view';
export var ShareElement = createComponent('taro-share-element');
export var Swiper = createComponent('taro-swiper');
export var SwiperItem = createComponent('taro-swiper-item', ['swiper-slide']);
export var View = createComponent('taro-view'); // 基础内容

export { default as Icon } from './icon';
export var Progress = createComponent('taro-progress', ['weui-progress']);
export var RichText = createComponent('taro-rich-text');
export { default as Text } from './text'; // 表单组件

export var Button = createComponent('taro-button');
export var Checkbox = createComponent('taro-checkbox', ['weui-cells_checkbox']);
export var CheckboxGroup = createComponent('taro-checkbox-group');
export var Editor = createComponent('taro-editor');
export var Form = createComponent('taro-form');
export var Input = createFormsComponent('taro-input', 'input');
export var KeyboardAccessory = createComponent('taro-keyboard-accessory');
export var Label = createComponent('taro-label');
export var Picker = createFormsComponent('taro-picker', 'change');
export var PickerView = createComponent('taro-picker-view');
export var PickerViewColumn = createComponent('taro-picker-view-column');
export var Radio = createComponent('taro-radio', ['weui-cells_checkbox']);
export var RadioGroup = createComponent('taro-radio-group', ['weui-cells_radiogroup']);
export var Slider = createFormsComponent('taro-slider', 'change', 'value', ['weui-slider-box']);
export var Switch = createFormsComponent('taro-switch', 'change', 'checked');
export var Textarea = createFormsComponent('taro-textarea', 'input'); // 导航

export var FunctionalPageNavigator = createComponent('taro-functional-page-navigator');
export var Navigator = createComponent('taro-navigator'); // 媒体组件

export var Audio = createComponent('taro-audio');
export var Camera = createComponent('taro-camera');
export { default as Image } from './image';
export var LivePlayer = createComponent('taro-live-player');
export var Video = createComponent('taro-video', ['taro-video-container']);
export var VoipRoom = createComponent('taro-voip-room'); // 地图

export var Map = createComponent('taro-map'); // 画布

export var Canvas = createComponent('taro-canvas'); // 开放能力

export var WebView = createComponent('taro-web-view');
export var Ad = createComponent('taro-ad');
export var AdCustom = createComponent('taro-ad-custom');
export var OfficialAccount = createComponent('taro-official-account');
export var OpenData = createComponent('taro-open-data'); // 导航栏

export var NavigationBar = createComponent('taro-navigation-bar'); // 页面属性配置节点

export var PageMeta = createComponent('taro-page-meta'); // 其他

export var Block = createComponent('taro-block');
export var CustomWrapper = createComponent('taro-custom-wrapper');
export var Slot = createComponent('taro-slot');
