// @tarojs/components/dist-h5/react/components/index.js
import { Fragment } from 'react';
import reactifyWc from '../utils/reactify-wc'; // 视图容器

export var CoverImage = reactifyWc('taro-cover-image-core');
export var CoverView = reactifyWc('taro-cover-view-core');
export var MatchMedia = reactifyWc('taro-match-media-core');
export var MovableArea = reactifyWc('taro-movable-area-core');
export var MovableView = reactifyWc('taro-movable-view-core');
export var PageContainer = reactifyWc('taro-page-container-core');
export var RootPortal = reactifyWc('taro-root-portal-core');
export var ScrollView = reactifyWc('taro-scroll-view-core');
export var ShareElement = reactifyWc('taro-share-element-core');
export var Swiper = reactifyWc('taro-swiper-core');
export var SwiperItem = reactifyWc('taro-swiper-item-core');
export var View = reactifyWc('taro-view-core'); // 基础内容

export var Icon = reactifyWc('taro-icon-core');
export var Progress = reactifyWc('taro-progress-core');
export var RichText = reactifyWc('taro-rich-text-core');
export var Text = reactifyWc('taro-text-core'); // 表单组件

export var Button = reactifyWc('taro-button-core');
export var Checkbox = reactifyWc('taro-checkbox-core');
export var CheckboxGroup = reactifyWc('taro-checkbox-group-core');
export var Editor = reactifyWc('taro-editor-core');
export var Form = reactifyWc('taro-form-core');
export { default as Input } from './input';
export var KeyboardAccessory = reactifyWc('taro-keyboard-accessory-core');
export var Label = reactifyWc('taro-label-core');
export var Picker = reactifyWc('taro-picker-core');
export var PickerView = reactifyWc('taro-picker-view-core');
export var PickerViewColumn = reactifyWc('taro-picker-view-column-core');
export var Radio = reactifyWc('taro-radio-core');
export var RadioGroup = reactifyWc('taro-radio-group-core');
export var Slider = reactifyWc('taro-slider-core');
export var Switch = reactifyWc('taro-switch-core');
export var Textarea = reactifyWc('taro-textarea-core'); // 导航

export var FunctionalPageNavigator = reactifyWc('taro-functional-page-navigator-core');
export var Navigator = reactifyWc('taro-navigator-core');
export var NavigationBar = reactifyWc('taro-navigation-bar-core'); // 媒体组件

export var Audio = reactifyWc('taro-audio-core');
export var Camera = reactifyWc('taro-camera-core');
export var Image = reactifyWc('taro-image-core');
export var LivePlayer = reactifyWc('taro-live-player-core');
export var LivePusher = reactifyWc('taro-live-pusher-core');
export var Video = reactifyWc('taro-video-core');
export var VoipRoom = reactifyWc('taro-voip-room-core'); // 地图

export var Map = reactifyWc('taro-map-core'); // 画布

export var Canvas = reactifyWc('taro-canvas-core'); // 开放能力

export var Ad = reactifyWc('taro-ad-core');
export var AdCustom = reactifyWc('taro-ad-custom-core');
export var OfficialAccount = reactifyWc('taro-official-account-core');
export var OpenData = reactifyWc('taro-open-data-core');
export var WebView = reactifyWc('taro-web-view-core'); // 页面属性配置节点

export var PageMeta = reactifyWc('taro-page-meta-core'); // 其他

export var Block = Fragment;
export var CustomWrapper = reactifyWc('taro-custom-wrapper-core');
